### DISCUSSION TITLE
Low Performance in Version 3

### DISCUSSION THREAD

**Discussion by luanfreitasdev** [livewire/livewire#6052] [7. Bugs (archived)] [answered]
### Livewire version

v3 (commit: 7808c12)

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

Hello!.

_(Keep calm, the title is the same. I come to discuss a problem and not to cause inconvenience 😃)_

I did a performance test comparing livewire v3 with livewire v2 using Powergrid. Read the important points below:

In powergrid, several @ifs, @elses, and @includes are used to do part of the work that simplifies the creation of a dynamic table for the developer. This described bug, I believe is related to how the dom is used in version 3.

Both tests were done without modifying the elements inside the table (divs, includes, and all the other stuff).

**👉  In the video below, we have livewire v2 with an average load time of 282 ms and 142 kb.**

https://github.com/livewire/livewire/assets/33601626/722f66ff-f504-4424-a9dd-852948687d95

**👉 And in this video, we have livewire v3 with an average load time of 2.75s and 172kb.**

https://github.com/livewire/livewire/assets/33601626/2cb4be30-d338-4367-9cb5-5a0a3e9a1a13

| Livewire Version | Size | Time |
|--------|--------|--------|
| v2 |  142 kb | 282 ms |
| v3 | 172kb | 2.75s | 

**Note**: In a simple listing (few elements in the dom, loading, and similar). I believe that maybe the bigger the livewire component is, the slower it will load in version 3.

Sorry but I don't know how to make a test fail in this regard to help fix it, let me know if I'm missing something or if this is a known issue due to changes.

Thanks!

**Comment by nuernbergerA** [comment, marked as answer]
can you do a test with `'inject_morph_markers' => false,` in your `config/livewire.php`?

**Reply by luanfreitasdev** [reply]
Thank you @nuernbergerA . It helped 50%. but still the difference is big

v3
![image](https://github.com/livewire/livewire/assets/33601626/0b2f1155-851e-4880-988d-cf7b6f9daac2)

v2
![image](https://github.com/livewire/livewire/assets/33601626/a54ac845-b0b8-46c2-a130-6d894e7fa1e4)

**Reply by luanfreitasdev** [reply]
After some more testing, this fixed it. Thanks!

**Reply by luanfreitasdev** [reply]
https://github.com/livewire/livewire/assets/33601626/a380c17c-6486-4d94-9711-949fa8a7091b

**Reply by julesgraus** [reply]
Hmmm.... So, setting ```inject_morph_markers ``` to false, makes it faster? Probable because it doesn't need to parse the templates via regex anymore then. But the reason ```inject_morph_markers```` exists is so that livewire can more easily detect the difference between a change and an addition (stated in the manual: https://livewire.laravel.com/docs/morphing#injecting-morph-markers).

But if it isn't able to easily detect the difference between a change and an addition, does it mean that bugs occur? If so, isn't  that's something that needs to  addressed?

I'm just trying to determine what needs to be done, if any.

**Reply by luanfreitasdev** [reply]
I did another test, now returning the value to 'true' with the last commit so far (_optimize: clear_) and it looks like it was similar to the last result. So it could be a problem with my environment or with the originally tested commit.

**Reply by joshhanley** [reply]
@julesgraus to answer your question, yes, there would potentially be morphing/ dom diffing bugs without it. But you can write views in a way that minimises it, which requires ensuring conditionals are wrapped in divs/elements and making use of `wire:key` on any root elements inside a conditional. That was how we had to do it in V2, so this should make it simpler.

I wonder if doing a view:cache will add the markers to the compile blade views? If so, then the potential performance hit would only be on the initial compilation of the blade view.

**Reply by luanfreitasdev** [reply]
@joshhanley In some cases when 'inject_morph_markers' => true I have this error:

- Uncaught TypeError: Cannot read properties of undefined (reading 'nodeType')

https://github.com/livewire/livewire/blob/main/dist/livewire.js#L7579

This fixes the problem, but I don't know what the consequences are
```js
 let isIf = (node) => node?.nodeType === 8 && node?.textContent === " __BLOCK__ ";
 let isEnd = (node) => node?.nodeType === 8 && node?.textContent === " __ENDBLOCK__ ";
```


https://github.com/livewire/livewire/assets/33601626/78204547-8d00-4dc5-b6a8-1a6532a33576

**Reply by calebporzio** [reply]
Hey @luanfreitasdev - I fixed that .nodeType bug - will be in the next release

**Reply by calebporzio** [reply]
Also, to @joshhanley's point - the injecting of the morph markers is only on Blade compile - so it shouldn't slow down production or refreshing the page - is that what you're finding as well?

**Reply by JoonD** [reply]
> After some more testing, this fixed it. Thanks!

greetings sir, I have similar problem. my livewire 3 response time more than 1s, I already read the other comment but I don't really understand. and I already tried to change "inject_morph_markers" to false and run artisan optimize:clear, artisan cache:clear. but I still got more than 1s response time

![image](https://github.com/user-attachments/assets/23235b4d-5ce0-4943-be61-202a88e4bb13)

**Reply by nuernbergerA** [reply]
@JoonD I assume you have a large public dataset can you post your livewire component PHP code

**Comment by calebporzio** [comment]
To clarify here, please provide the following information:

A) Is the slowness from the network request? or in JavaScript?
B) If setting "inject_morph_markers" to false makes it fast again, is it back to V2 speed?
C) If you keep inject_morph_markers as false, does simply refreshing the page fix the problem? What about `artisan view:cache`?

Thanks

**Reply by luanfreitasdev** [reply]
a - after a few more tests, I realized that the problem was compiling the assets of the entire project, as soon as I ran artisan optimize:clear, it improved a lot.
b - was about 10% slower than version 3, but still much faster than version 2.
c - needed to clear cache
